sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+ *		Luo Xionghu <xionghu.luo@intel.com>
+ *		Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *		Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+/*
+ * virt IO FE driver
+ *
+ * The SOF driver thinks this driver is another audio DSP, however the calls
+ * made by the SOF driver core do not directly go to HW, but over a virtIO
+ * message Q to the virtIO BE driver.
+ *
+ * The virtIO message Q uses the *exact* same IPC structures as we currently
+ * use in the mailbox.
+ *
+ * The mailbox IO and TX/RX msg functions below do IO on the virt IO Q.
+ */
+
+#include <linux/completion.h>
+#include <linux/device.h>
+#include <linux/firmware.h>
+#include <linux/interrupt.h>
+#include <linux/module.h>
+#include <linux/scatterlist.h>
+#include <linux/virtio.h>
+#include <linux/virtio_config.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_ring.h>
+#include <sound/sof.h>
+#include <sound/sof/virtio.h>
+
+#include "ops.h"
+#include "sof-priv.h"
+
+#define SOF_COMPONENT_NAME	"sof-audio-component"
+#define DATA_TIMEOUT_MS		600
+
+static const char *const sof_vq_names[SOF_VIRTIO_NUM_OF_VQS] = {
+	SOF_VIRTIO_IPC_CMD_VQ_NAME,
+	SOF_VIRTIO_IPC_PSN_VQ_NAME,
+	SOF_VIRTIO_DATA_VQ_NAME,
+};
+
+struct sof_vfe {
+	struct snd_sof_dev *sdev;
+
+	/* IPC cmd from frontend to backend */
+	struct virtqueue *ipc_cmd_vq;
+
+	/* IPC position update from backend to frontend */
+	struct virtqueue *ipc_psn_vq;
+
+	/* audio data in both directions */
+	struct virtqueue *data_vq;
+
+	/* position update work */
+	struct work_struct posn_update_work;
+
+	/* current pending cmd message */
+	struct snd_sof_ipc_msg *msg;
+
+	/* current and pending notification */
+	struct snd_sof_ipc_msg *not;
+	struct sof_ipc_stream_posn posn;
+
+	struct sof_vfe_ipc_tplg_resp tplg;
+
+	struct completion completion;
+
+	/* A shared capture / playback virtual queue data buffer */
+	union {
+		struct dsp_sof_data_req data_req;
+		struct dsp_sof_data_resp data_resp;
+	};
+
+	/* Headers, used as a playback response or capture request */
+	union {
+		u8 hdr_req[HDR_SIZE_REQ];
+		u8 hdr_resp[HDR_SIZE_RESP];
+	};
+};
+
+/*
+ * IPC Firmware ready.
+ */
+static int sof_vfe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
+{
+	return 0;
+};
+
+/* used to send IPC to BE */
+static int sof_vfe_send_msg(struct snd_sof_dev *sdev,
+			    struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	int ret;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
+	size_t msg_size = msg->msg_size;
+	void *msg_data = msg->msg_data;
+
+	switch (msg->header) {
+	case SOF_IPC_GLB_STREAM_MSG | SOF_IPC_STREAM_TRIG_START:
+		sg_init_one(&sg_in, &vfe->posn,
+			    sizeof(struct sof_ipc_stream_posn));
+		ret = virtqueue_add_inbuf(vfe->ipc_psn_vq,
+					  &sg_in, 1, &vfe->posn, GFP_KERNEL);
+		if (ret < 0) {
+			dev_err(sdev->dev, "%s(): failed %d to add a buffer\n",
+				__func__, ret);
+			return ret;
+		}
+
+		virtqueue_kick(vfe->ipc_psn_vq);
+
+		break;
+	}
```

Comment:
default case?

Reply:
@plbossart a default case of "do nothing" like
```
switch (x) {
case 0:
...
break;
default:
break;
}
```
isn't required in C. Only if `x` of af on `enum` type and not all `enum` values are handled, the compiler will complain. So, only in that case you have to either handle each `enum` value explicitly of have a `default` fallback, even if it does nothing.

Reply:
@lyakh Coding style rules demand that you still need to specify the default case unless you have exhausted all possibilities, even if it's a do-nothing default. So I'm with Pierre here.

Reply:
switch (msg->header) 

this is a u32, how can we test only one value?

Reply:
@paulstelian97 sorry, I'm unaware of this coding style "requirement" and I don't find it in `Documentation/process/coding-style.rst`. Could you provide a reference?

@plbossart sorry, I don't see a problem? That's a perfectly valid `switch`-`case` C construct, as far as I know.